yjit_codegen.c

```diff
@@ -1791,31 +1793,56 @@ gen_return_branch(codeblock_t* cb, uint8_t* target0, uint8_t* target1, uint8_t s
 }
 
 static codegen_status_t
-gen_oswb_iseq(jitstate_t *jit, ctx_t *ctx, const struct rb_callinfo *ci, const rb_callable_method_entry_t *cme, int32_t argc)
+gen_oswb_iseq(jitstate_t *jit, ctx_t *ctx, const struct rb_callinfo *ci, const rb_callable_method_entry_t *cme, const int32_t argc)
 {
     const rb_iseq_t *iseq = def_iseq_ptr(cme->def);
-    const VALUE* start_pc = iseq->body->iseq_encoded;
-    int num_params = iseq->body->param.size;
-    int num_locals = iseq->body->local_table_size - num_params;
 
-    if (num_params != argc) {
-        GEN_COUNTER_INC(cb, oswb_iseq_argc_mismatch);
+    if (vm_ci_flag(ci) & VM_CALL_TAILCALL) {
+        // We can't handle tailcalls
+        GEN_COUNTER_INC(cb, oswb_iseq_tailcall);
         return YJIT_CANT_COMPILE;
     }
 
-    if (!rb_simple_iseq_p(iseq)) {
-        // Only handle iseqs that have simple parameters.
-        // See vm_callee_setup_arg().
-        GEN_COUNTER_INC(cb, oswb_iseq_not_simple);
-        return YJIT_CANT_COMPILE;
+    // Arity handling and optional parameter setup
+    int num_params = iseq->body->param.size;
+    uint32_t start_pc_offset = 0;
+    if (rb_simple_iseq_p(iseq)) {
+        if (num_params != argc) {
+            GEN_COUNTER_INC(cb, oswb_iseq_arity_error);
+            return YJIT_CANT_COMPILE;
+        }
     }
+    else if (rb_iseq_only_optparam_p(iseq)) {
+        // We follow the logic of vm_call_iseq_setup_normal_opt_start()
+        // and these are the preconditions required for using that fast path.
+        RUBY_ASSERT(vm_ci_markable(ci) && ((vm_ci_flag(ci) &
+                        (VM_CALL_KW_SPLAT | VM_CALL_KWARG | VM_CALL_ARGS_SPLAT)) == 0));
+
+        const int lead_num = iseq->body->param.lead_num;
+        const int opt = argc - lead_num;
+        const int opt_num = iseq->body->param.opt_num;
```

Comment:
IMO `opt` and `opt_num` need better names. Maybe better commenting. It's not clear to me what they represent. Not sure what is `lead_num` either. Is that non-optional params?